packages/plugin-table/src/TableDomParser.ts

```diff
@@ -1,20 +1,129 @@
 import { AbstractParser } from '../../core/src/AbstractParser';
 import { DomParsingEngine } from '../../plugin-dom/src/DomParsingEngine';
 import { TableNode } from './TableNode';
+import { TableRowNode } from './TableRowNode';
+import { TableCellNode } from './TableCellNode';
 
 export class TableDomParser extends AbstractParser<Node> {
     static id = 'dom';
     engine: DomParsingEngine;
 
-    predicate = (item: Node): boolean => {
-        return item instanceof Element && item.tagName === 'TABLE';
+    predicate = (item: Node): item is HTMLTableElement => {
+        return item.nodeName === 'TABLE';
     };
 
-    async parse(item: Element): Promise<TableNode[]> {
+    /**
+     * Parse a table node.
+     *
+     * @param item
+     */
+    async parse(item: HTMLTableElement): Promise<TableNode[]> {
+        // Parse the table itself and its attributes.
         const table = new TableNode();
         table.attributes = this.engine.parseAttributes(item);
-        const nodes = await this.engine.parse(...item.childNodes);
-        table.append(...nodes);
+
+        // Parse the contents of the table.
+        const children = await this.engine.parse(...item.childNodes);
+
+        // Build the grid.
+        const dimensions = this._getTableDimensions(item);
+        const parsedRows = children.filter(row => row.is(TableRowNode)) as TableRowNode[];
+        const grid = this._createTableGrid(dimensions, parsedRows);
+
+        // Append the cells to the rows.
+        const rows = new Array<TableRowNode>(dimensions[0]);
+        for (let rowIndex = 0; rowIndex < grid.length; rowIndex += 1) {
+            rows[rowIndex] = parsedRows[rowIndex];
+            const cells = grid[rowIndex];
+            let row = rows[rowIndex];
+            if (!row) {
+                row = new TableRowNode();
+            }
+            row.append(...cells);
+        }
+
+        // Append the rows to the table.
+        table.append(...rows);
         return [table];
     }
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * Return a tuple with the row length and the column length of the given DOM
+     * table element.
+     *
+     * @param domTable
+     */
+    _getTableDimensions(domTable: HTMLTableElement): [number, number] {
+        const domRows: HTMLTableRowElement[] = [];
+        for (const domChild of domTable.querySelectorAll('tr')) {
+            domRows.push(domChild);
```

Comment:
Can't the result of `querySelectorAll` be transformed into an array rather than pushing it manually ?